app/js/lib/equation_editor/grammar.ne

```diff
@@ -0,0 +1,286 @@
+    let rx = new RegExp(_.keys(greekLetterMap).join('|'), 'g')
+    let parts = d[0].text.replace(rx, (v) => greekLetterMap[v] || v).split('_')
+    let topChain = _processChainOfLetters(parts[0])
+    if (parts.length > 1) {
+        let chain = _processChainOfLetters(parts[1])
+        let r = _findRightmost(topChain)
+        r.children['subscript'] = chain
+    }
+    return topChain
+}
+
+const processIdentifierModified = (d) => {
+    let rx = new RegExp(_.keys(greekLetterMap).join('|'), 'g')
+    let parts = d[0].text.split('_')
+    let topChain = _processChainOfLetters(parts[0].replace(rx, (v) => greekLetterMap[v] || v))
+    let r = _findRightmost(topChain)
+    r.properties['modifier'] = parts[1]
+    return topChain
+}
+
+const processNumber = (d) => {
+    return { type: 'Num', properties: { significand: d[0].text }, children: {} }
+}
+
+const processDerivative = (d) => {
+    let numerator = {
+        type: 'Differential',
+        properties: { letter: 'd' },
+        children: {
+            argument: d[3]
+        }
+    }
+    let denList = d[7].reduce((a, e) => {
+        if (a.length == 0) {
+            return [{ object: e, order: 1 }]
+        } else {
+            let last = a[a.length-1]
+            if (e.type === 'Num') {
+                last.order = parseInt(e.properties.significand)
+                return [...a.slice(0, a.length-1), last]
+            } else if (_.isEqual(e, last.object)) {
+                last.order = last.order + 1
+                return [...a.slice(0, a.length-1), last]
+            } else {
+                return [...a, { object: e, order: 1 }]
+            }
+        }
+    }, []).map(e => {
+        let differential = {
+            type: 'Differential',
+            properties: { letter: 'd' },
+            children: {
+                argument: e.object
+            }
+        }
+        if (e.order > 1) {
+            differential.children['order'] = {
+                type: 'Num',
+                properties: { significand: `${e.order}` },
+                children: {}
+            }
+        }
+        return differential
+    })
+    let order = denList.reduce( (a, e) => a + (parseInt(e.children.order ? e.children.order.properties.significand : "1")), 0)
+    if (order > 1) {
+        numerator.children['order'] = {
+            type: 'Num',
+            properties: { significand: `${order}` },
+            children: {}
+        }
+    }
+    let denominator = _.reduceRight(denList, (a, c) => {
+        c.children['right'] = a
+        return c
+    })
+    return {
+        type: 'Derivative',
+        children: {
+            numerator,
+            denominator
+        }
+    }
+}
+%}
+
+@lexer lexer
+
+### Behold, the Grammar!
+
+main -> _ AS _                                            {% processMain %}
+      | _ AS _ %Rel _ AS _                                {% processRelation %}
+
+P ->               %Lparen _ AS _                 %Rparen {% processBrackets %}
```

Comment:
Why is the spacing so strange here? The [chemistry checker grammar](https://github.com/jsharkey13/chemistry-checker/blob/master/src/main/cup/ChemistryParser.cup) had to be in a similar form but looks much more readable than this does!